### ISSUE TITLE
Feature request: per-project Python interpreter

### ISSUE THREAD

**Issue by akdor1154** [spyder-ide/spyder#11362]
Related issues:
#9804 (project settings, you guys are considering Conda envs which is a similar usecase)
#1014 (exactly this, imported from google code and seemingly abandoned)

Currently interpreters are a global setting. It would be great to have a Python interpreter selectable as a project-level preference instead.

In #9804 it looks like you are considering allowing project-level Conda, but I don't use Conda. I just use a virtualenv per project, managed by `poetry`. This means I just want to set every project's interpreter to `./.venv/bin/python`, relative to the project root.

**Comment by LeoWelter** [comment]
may I say "me too". I was kind of expecting this would be in. As new user of Spyder i'm kind of disappointed since virtual env support is in pycharm and eclipse for ages now.

**Comment by mhamiltonj** [comment]
The lack of this feature means the team I work with is now migrating to PyCharm, despite the fact it has drawbacks for data science work. Keeping multiple team members' enviroments consistent with conda was just too cumbersome, so we switched to pipenv, only to find that dealing with different envirmoments in spyder requires manually changing the interpreter path in settings every time you switch project. Given that we all work on several projects, and often need to switch between them quickly, this is just not a usable workflow.

**Comment by ccordoba12** [comment]
Sorry to hear that. Unless you and your team are willing to fund this feature, it'll come in Spyder 5 (to be released next year).

**Comment by svdHero** [comment]
@mhamiltonj we have exactly the same problem in our company. Switched to pipenv and wanted to migrate from PyCharm to Spyder, which is especially great for our physicists, because Spyder has such a Matlab-Feel to it.
The only obstacle remaining is the need to manually change the interpreter path many times a day. So annoying and error-prone.

Spyder is great, but it definetely needs this feature in order to be of any significance in a professional work environment (pun intended :wink: ).

**Comment by ccordoba12** [comment]
@svdHero, I said in my comment above:

> Unless you and your team are willing to fund this feature, it'll come in Spyder 5 (to be released next year).

Unfortunately, this is not true anymore. Furthermore, we don't have plans to implement this now because it's been requested very little.

So, if this is important for your company, please consider to fund this work. You can contact me (my email address is in my profile) to discuss about it.

**Comment by PackElend** [comment]
there is a  similar request, see #4223

**Comment by svdHero** [comment]
@ccordoba12 Unfortunately, funding is not possible, since we are a rather small electronics manufacturing company and the main focus is not Python. So management has no interest in investing into Python toolchain. Seems as if I will have to go back to VSCode until this feature makes it into Spyder.

I am sorry. I'd love our company to support open source projects. 😔

**Comment by ccordoba12** [comment]
@svdHero, sorry to hear that. Unfortunately, we are a small team and this is kind of a big feature. We'll try to implement it for Spyder 6, to be released at the end of this year or the beginning of 2022.

**Comment by mrclary** [comment]
@ccordoba12, Since Spyder projects already create a .spyproject directory containing project configuration settings, it seems that the core configuration paradigm already exists and the bulk of the development work would just be implementing a project configuration interface. I imagine a Project Configuration window that would be something like a stripped down version of Spyder's preferences window, with the ability to select an interpreter (and set code style linting, and ...).

This, of course, would be quite a bit of work; but does this resemble what you were thinking could be included for Spyder 5?

**Comment by ccordoba12** [comment]
Yes, that's basically the idea. We have a proposal for that dialog on issue #9804.

**Comment by frafra** [comment]
Wouldn't it be enough in some cases to add a command line option to use a specific python executable? Something like `--use-current-python-interpreter` and or `--python-intepreter PATH`?

It would be pretty easy just running `spyder -p . --use-current-python-interpreter`.

**Comment by trianta2** [comment]
I've been searching around for this feature request. I just installed Spyder 5.1.5 and loving the improvements over Spyder 3. Kind of surprised that project-specific metadata such as the Python interpreter isn't available. Looking forward to when this feature is added.

**Comment by ccordoba12** [comment]
This feature hasn't been implemented yet, sorry.

**Comment by EvenGu** [comment]
I'd like to request this feature as well. Once I create a spyder project, I would expect to be able to specify and save the interpreter associated so that I don't need to change it every time I work on another project in a different environment.

**Comment by invokermain** [comment]
This is a sorely needed feature, per project venv is best practice, using shared venv across many projects is not great and leads to many bad habits (one venv for everything, no per project dependency specification etc)

**Comment by ccordoba12** [comment]
We'll try to implement this feature for Spyder 6, but the problem is that offering the ability to associate an environment with a project also intuitively demands a graphical way to manage packages in that env. And that's simply a much harder endeavor for which we don't enough resources at the moment.

**Comment by jonzhaopeng** [comment]
We also need this feature, but we don't have money to donate. Maybe  We have engineers to write code

**Comment by ba-tno** [comment]
I encountered this issue as well and have created a workaround [in this gist](https://gist.github.com/ba-tno/0c353b58b4eb8c4430e8aad9bd3a3402) to deal with this in the meantime. Perhaps it's useful for others.

**Comment by wilecoyote2015** [comment]
We also need this feature. I've considered migrating from PyCharm to Spyder for scientific work several times in the last years, but this, along with missing PyCham-Like run-configurations, is always a showstopper, unfortunately.

**Comment by AZippelius** [comment]
I also really want this feature, and the lack makes working with multiple projects are real pain.

**Comment by ba-tno** [comment]
@AZippelius I think "a real pain" is exagerating a bit, but if you're really hindered by this you can use [the gist I linked to earlier](https://gist.github.com/ba-tno/0c353b58b4eb8c4430e8aad9bd3a3402) to work around this issue (on Windows).

**Comment by AZippelius** [comment]
@ba-tno As far as I can tell, that automatically selects the interpreter when starting Spyder, but not when switching projects inside spyder. So that would require a restart each time, which is only a marginal improvment.

**Comment by callegar** [comment]
@ccordoba12  > We'll try to implement this feature for Spyder 6, but the problem is that offering the ability to associate an environment with a project also intuitively demands a graphical way to manage packages in the env

Please, don't let this delay this useful feature. If I understand it correctly, as soon as the console in spyder gets associated with a virtual environment you already get a way to manage packages in the env from spyder almost for free: running `pip` in the console.

**Comment by gwerbin** [comment]
> @AZippelius I think "a real pain" is exagerating a bit, but if you're really hindered by this you can use [the gist I linked to earlier](https://gist.github.com/ba-tno/0c353b58b4eb8c4430e8aad9bd3a3402) to work around this issue (on Windows).

I would say it is indeed "a real pain" for _some_ workflows (mine) but not others (yours). For example, not having this feature is basically a dealbreaker for me, and it's been the \#1 thing keeping me away from using Spyder over the years (RStudio also lacks or lacked this feature for many years, and similarly I basically had to stop using it). I would love to see this implemented ASAP, because I do very much want to use Spyder.

I also strongly agree with @callegar in that you do not need GUI package management in order to make per-project environment/interpreter association useful, and that I hope the legitimate desire for the former does not delay the latter.

**Comment by 01tot10** [comment]
I'm in the process of trying to have my projects contained in separate (conda/mamba) environments, and ran into this limitation today. Wanted to give my thumbs up for the development team for considering this as a future feature!

**Comment by mrclary** [comment]
This feature is planned for Spyder 6.0 release later this year.

**Comment by yuval-harpaz** [comment]
> This feature is planned for Spyder 6.0 release later this year.

Any updates? I am using Spyder 6.0.2 and can't see preferences per project. After PyCharm disaster (upgrade to Ubuntu 24.04 did not go well) I returned to Spyder but the lack of this feature is a major issue for me

**Comment by ccordoba12** [comment]
@yuval-harpaz, sorry, we haven't had time to work on this. But it's really high in our todo list for 6.1, to be released in the first half of next year.

**Comment by VaibhavKrMishr** [comment]
@ccordoba12 
Hello!! Sir! is this still being worked on? Would love to help if not.

I hit this constantly — I use uv for project envs, and every time I switch projects in Spyder I have to manually deactivate conda and repoint the interpreter setting. Kills the flow compared to tools that just auto-detect the local venv, like VS Code — but my course requires Spyder specifically, so switching isn't an option for me.
